eskip/eskip.go

```diff
@@ -320,3 +324,119 @@ func GenerateIfNeeded(existingId string) string {
 	id = routeIdRx.ReplaceAllString(id, "x")
 	return "route" + id
 }
+
+func marshalJsonPredicates(r *Route) []*Predicate {
+	rjf := make([]*Predicate, 0)
+
+	if r.Method != "" {
+		rjf = append(rjf, &Predicate{
+			Name: "Method",
+			Args: []interface{}{r.Method},
+		})
+	}
+
+	if r.Path != "" {
+		rjf = append(rjf, &Predicate{
+			Name: "Path",
+			Args: []interface{}{r.Path},
+		})
+	}
+
+	for _, h := range r.HostRegexps {
+		rjf = append(rjf, &Predicate{
+			Name: "HostRegexp",
+			Args: []interface{}{h},
+		})
+	}
+
+	for _, p := range r.PathRegexps {
+		rjf = append(rjf, &Predicate{
+			Name: "PathRegexp",
+			Args: []interface{}{p},
+		})
+	}
+
+	for k, v := range r.Headers {
+		rjf = append(rjf, &Predicate{
+			Name: "Header",
+			Args: []interface{}{k, v},
+		})
+	}
+
+	for k, list := range r.HeaderRegexps {
+		for _, v := range list {
+			rjf = append(rjf, &Predicate{
+				Name: "HeaderRegexp",
+				Args: []interface{}{k, v},
+			})
+		}
+	}
+
+	for _, p := range r.Predicates {
+		rjf = append(rjf, p)
+	}
+
+	return rjf
+}
+
+func (f *Filter) MarshalJSON() ([]byte, error) {
+	args := f.Args
+	if args == nil {
+		args = []interface{}{}
+	}
+
+	return json.Marshal(&struct {
+		Name string        `json:"name"`
+		Args []interface{} `json:"args"`
+	}{
+		Name: f.Name,
+		Args: args,
+	})
+}
+
+func (f *Predicate) MarshalJSON() ([]byte, error) {
```

Comment:
to avoid duplication, there could be a s.g. like:

```
func marhsalNameArgs(name string, args []interface{}) ([]byte, error) {
    	if args == nil {
		args = []interface{}{}
	}

	return json.Marshal(&struct {
		Name string        `json:"name"`
		Args []interface{} `json:"args"`
	}{
		Name: name,
		Args: args,
	})
}
```

and called from both the predicate and the filter.

+1 for the anonymous struct